[RFXCOM] Fixes to shutdown and packet reading thread

mjagdis opened the pull request:
We MUST set read timeouts on the input streams because the reads
themselves are not interruptible. We SHOULD wait for the read thread
to handle the interrupt and exit before completing the shutdown.

The packet reading loop SHOULD NOT handle an interrupt and shutdown
in the middle of a packet - only between packets. In the unlikely
(but possible) event that the RFXCOM is unplugged mid-packet a
timeout is used while reading packet data and the connection
abandoned if it is reached.

Closes #2100

Signed-off-by: Mike Jagdis <mjagdis@eris-associates.co.uk> (github: mjagdis)

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,39 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
 
-                    for (int i = 0; i < len; i++) {
+                if (in.read(buf, 0, 1) > 0 && buf[0] > 0) {
+                    int i = 1;
```
We could maybe also extra the reading of the first byte outside of the if, otherwise the if does so much..

mjagdis replied:
i is a index to the buffer. This loop is a pretty standard C idiom however I've increased the verbiage a bit. Personally I think that makes it _harder_ to read but that's because I write a lot more C than Java :-)

mjagdis replied:
Ok

mjagdis commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,39 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
-                    logger.trace("Received data (len={}): {}", len, DatatypeConverter.printHexBinary(logData));
+        while (!Thread.interrupted()) {
+            try {
+                byte[] buf = new byte[Byte.MAX_VALUE];
```
Good point!

mjagdis commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,39 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
-                            if (index == msgLen) {
-                                sendMessage(dataBuffer, msgLen);
+                    while (i <= buf[0]) {
+                        int n = in.read(buf, i, buf[0] - i + 1);
```
Ok

mjagdis commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,39 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
-                            }
+                        if (n > 0) {
+                            i += n;
+                        } else if (intCount++ == 4) {
```
It's the count of how many times we took a read timeout while in the middle of reading a packet.

mjagdis commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,39 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
+                        if (n > 0) {
+                            i += n;
+                        } else if (intCount++ == 4) {
+                            throw new IOException("Timeout during packet read");
```
Yes we should. We should not expect to see _any_ read timeouts in the middle of a packet really. But if they _do_ occur it means the RFXCOM is not doing what it should be doing and we cannot just ignore it. That would leave the bridge ONLINE but simply not giving us the input it should which means your lights stop working. The least we can do is throw an error up the stack to trigger a reconnect and reset.

martinvw replied:
Ok but normally it's not good practice to throw core exceptions, use exceptions for flow control (you are throwing and catching in the same method).

Maybe you can just do:

```
connector.sendErrorToListeners("Timeout during packet read");
return;
```

mjagdis commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,39 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
-                    } catch (InterruptedException ignore) {
-                    }
+
+                    connector.sendMsgToListeners(Arrays.copyOfRange(buf, 0, buf[0] + 1));
```
No. It just adds nothing useful. All it did was copy the packet out of the buffer and hand it to the connector's method. We can do that ourselves without the bloat.

martinvw replied:
I would not call, using additional (short) methods, call bloat. But your call is indeed not much longer than the method call so do agree for this case :-)

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/handler/RFXComHandler.java:
```diff
@@ -165,108 +166,112 @@ public void onDeviceMessageReceived(ThingUID bridge, RFXComMessage message) {
                 }
             }
-        } catch (Exception e) {
+        } catch (RFXComException e) {
```
Can there be errors on this level, are they not already caught on the inner catch?

mjagdis replied:
That was leakage from the other PR but yes, both getDeviceId and getSupportedInputValueSelectors can throw this exception.

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComBaseMessage.java:
```diff
@@ -147,4 +149,9 @@ public String toString() {
     }
+
+    @Override
+    public Command convertToCommand(RFXComValueSelector valueSelector) throws RFXComException {
```
This branch suddenly contains both changes 👎

mjagdis replied:
Oops!

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,50 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
+                    int bufferIndex = 1;
+                    int readTimeoutCount = 1;
+
+                    while (bufferIndex <= buf[0]) {
```
I think this buf[0] should be packetLength

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,39 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
+            } catch (IOException ioe) {
+                connector.sendErrorToListeners(ioe.getMessage());
+                try {
+                    Thread.sleep(100);
```
👍

martinvw commented:
This looks way better, can you undo combining both PR's in this one?

If you are finished I would like to take a test ride with your code later on this week or weekend. I hope some others can join in as well.

martinvw commented:
I also had some responses in the collapsed / outdated parts which you might have missed.

---

Ok but normally it's not good practice to throw core exceptions and/or use exceptions for flow control (you are throwing and catching in the same method).

Maybe you can just do:

```
connector.sendErrorToListeners("Timeout during packet read");
return;
```

---

Are you aware that intCount++ returns the value before incrementing?

so:

```java
int i = 0;
int t = i++;
// t = 0, i = 1
```
---
I would not call, using additional (short) methods, bloat. But your call on itself is indeed not much longer than the method call to the external method, so do completely agree for this case :-)

And finally:

If you have a jar ready it would be great if you could upload it. Otherwise I need to find some time to check out your repo and build that one.

mjagdis commented:
We could send an error but then what? We can assume that whatever happened the current packet is toast. We don't want to simply return because we don't want to make assumptions about how the error is to be handled. We do want to exit the packet reading loop - that packet is dead and we want to be back in the outer loop where we watch for interrupt - but we can't just break out of the packet loop or we'll send a partially filled (or fully filled but not with the right data...) packet up the stack. We _could_ do it without the local throw-catch but only at the expense of complicating the code by mixing the normal case with error handling conditions - exactly what exceptions are there to avoid.

Yes, post-increment behaves as specified and documented. There is no reason to make this condition dependent on waiting for the result of the calculation.

I dunno. I'd call replacing one call with another call at the expense of adding several more lines and hoping the optimizer is smart enough, bloat :-)

Here's a jar with these three PRs merged. It's not quite what I'm running now since I've got some Linux serial connection discovery/PnP in the pipeline but...

[rfxcom-2100+2102+2125.zip](https://github.com/openhab/openhab2-addons/files/904071/rfxcom-2100.2102.2125.zip)

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,50 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
+                        if (bytesRead > 0) {
+                            bufferIndex += bytesRead;
+                        } else if (readTimeoutCount++ == MAX_READ_TIMEOUTS) {
+                            throw new IOException("Timeout during packet read");
```
Ok, note that I'm referring to this line, this exception is caught on line 74 and does not terminate the while loops. In fact no IOException while ever stop the while loop. And it also does not ever seem to force a disconnect or reconnect?

What happens if I pull out and wait eg 10 minutes and put it back in? I think to remember that this used to work in OH2

martinvw replied:
So replacing this with 
```
connector.sendErrorToListeners("Timeout during packet read");
continue;
```

Would give the same result as the current option with throwing an exception.

mjagdis replied:
A continue would be fatal. The inner packet-reading loop (deliberately) does not check for interrupts and read is NOT an interruptible (which was a major problem with the original). If you continue that loop then a stalled (as opposed to I/O erroring) rfxcom becomes non-recoverable. The throw jumps over the handling for the completed packet and into the error handler for the outer loop. This sends the error upstream and continues in the outer loop waiting for a new packet length header. This inter-packet loop DOES check for interrupts and handles shutdown gracefully.

Now sending the error upstream feeds errorOccurred in RFXComBridgeHandler, which logs it and sets the bridge offline. If a packet is then received successfully (remember the stream reader is waiting in hope) then the bridge handler's MessageListener.packetReceived will set it back online. Alternatively the bridge handler has a "check" job that fires every minute. If that job fires before a packet is received it sees the bridge is offline and calls connect(). Connect() disconnects any existing connector (part of which involves interrupting the stream read to tell it to go away) and attempts to connect a new one. If successful we're back online again. Otherwise we just keep trying every minute. Similarly taking an error during sending a message takes the bridge offline and relies on the same packet-received-or-check-job to fix things.

martinvw replied:
> A continue would be fatal.
Correct, I missed the second nested loop. So what we actually need to do to avoid throwing and catching an exception in the same method is breaking the inner loop and continue-the outer loop. Java has ways for that but those are a little uncommon. 
_The most proper way is imho using more methods, I think we learned programming from different persons because I always like more methods and like them between 2 & 10 lines._

Thanks for the complete explanation of the error handling 👍

mjagdis replied:
In the Linux world of C we'd write the normal case then use a goto to jump out to the error handler :-). Hmmm... Java has gotos though. They're not _called_ goto, they're disguised as labelled breaks and continues. We can replace the inner throw with a sendError and continue the outer loop...

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,50 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
+                connector.sendErrorToListeners(ioe.getMessage());
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException ie) {
```
If we receive this exception, the interrupted flag was cleared and we need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/

martinvw replied:
The same might hold for the other catch `InterruptedException`

mjagdis replied:
Good point. There should be a break in there. We've been interrupted so we want out.

martinvw commented:
Except for the thing about the interupted exeption it looks perfect now. Thanks for the all the extensive explanations.

I just started some testing on my dev-machine.

martinvw commented:
Retrying every 100ms does explode the system out it seems that an underlying lib is doing some ugly error output; So looking at your explanation above, it might try it up to 600 times before forcing a reconnect...

```
Error 0x5 at src/windows/termios.c(2725): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(512): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(2725): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(512): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(2725): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(512): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(2725): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(512): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(2725): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(512): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(2725): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(512): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(2725): Toegang geweigerd.


Error 0x5 at src/windows/termios.c(512): Toegang geweigerd.
```

Recovery works great

martinvw commented:
in fact I received 19101 errors with 
> Error 0x5 at src/windows/termios.c(512): Toegang geweigerd.

And 19100 with:
> Error 0x5 at src/windows/termios.c(2725): Toegang geweigerd.

```
2017-04-07 13:03:42.450 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'rfxcom:bridge:238adf67' updated: ONLINE
... the 19100 + 19101 error ...
2017-04-07 13:04:25.712 [DEBUG] [.b.r.i.c.RFXComSerialConnector:85   ] - Serial port event listener stopped
```

This does not seem to match what I thought your code should do any idea what might be happening.

Also the load seriously increases during that period, I can try to make a proper graph of it if you cannot reproduce it.

mjagdis commented:
Ah. Windows. I use Linux myself so we just need someone with OSX. Anyway, when a USB device disappears under Windows attempts to use an existing descriptor are failed with an access denied error (makes sense to someone obviously...). It looks like the native library (D2XX or RXTX) is "helpfully" logging that before converting to an IOException. Thanks. Anyway, like I said there's no recovery from a disappearing USB device without reconnecting on reinsertion so we might as well make an assumption about how the problem is going to be handled and end the reader after the sendError. This is fine because all the message going up the stack through sendError does is trigger an offline of the bridge and we know that in this case the only way to get it back is to wait for the check job to initiate a reconnect. I've added a new commit. If it looks good I'll squash it before it gets merged but for now you probably need it separate.

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,48 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
-            logger.error("Reading from serial port failed", e);
-            connector.sendErrorToListeners(e.getMessage());
+        } catch (IOException ioe) {
+            connector.sendErrorToListeners(ioe.getMessage());
```
Did you remove the thread sleep here? So it will fire even harder?

martinvw replied:
Sorry, having bad sleep with a young baby, so my reading fails some time :(

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java:
```diff
@@ -37,89 +36,48 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
+        // packet are ignored but if too many occur we take it as meaning
+        // the RFXCOM has become missing presumed dead.
+        try {
+start_packet:
```
maybe some auto-formating here?

martinvw commented:
I have some OSX around too, I can try that later on

martinvw commented:
And you can add yourself to the list of authors of this file since this is (almost) a complete rewrite :-)

PS: when checking out locally I saw that StreamReader now seems to have two unused imports.

martinvw commented:
Another test with the new code, most likely it is a bug in the library, it's still freaking out... and giving 1000> errors :-(

The errors even keep coming if I pause the Java application via the debuger

martinvw commented:
Works on Mac Os too, including stop/start while sending.

If you can fix the unused imports and small layout thing its in my opinion ready to be merged.

Thanks for all the clear explanations and your work in improving this binding!

mjagdis commented:
Ok. I've updated, squashed and pushed. I'm still not happy with the error spewing or the CPU consumption. I'm sure we can do better, but not in this PR.

lolodomo commented:
What is the change in CPU consumption ?
Is the binding still usable with a slow CPU hardware like a RPI ?
If these changes are working well, I will apply them to my Powermax binding.

martinvw commented:
@lolodomo the CPU consumption did change once you unplug the device but I did not compare it with the old situation. So it might have already been the case... Some in-between version of this PR reattempted the reading every 100 ms, but maybe that was not even the cause.

The final version should not have any relevant change in runtime cpu usage only the enabling of the timeouts could have a small impact.

lolodomo commented:
Ok

Doing a search, disableReceiveTimeout is now only used by 1 OH2 binding: the DSC Alarm binding.

I can guess that the DSC alarm binding has the issue that the RFXCOM binding had. It could make sense to fix it too. Something to check with the binding author.

disableReceiveTimeout is used by 7 OH1 bindings including RFXCOM, and DSC alarm bindings.

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComTcpConnector.java:
```diff
@@ -39,6 +39,7 @@
     public void connect(RFXComBridgeConfiguration device) throws IOException {
         logger.info("Connecting to RFXCOM at {}:{} over TCP/IP", device.host, device.port);
         socket = new Socket(device.host, device.port);
+        socket.setSoTimeout(100); // In ms. Small values mean faster shutdown but more cpu usage.
```
It seems to work different from what was assumed, if you set the [ssoTimeout](http://docs.oracle.com/javase/7/docs/api/java/net/Socket.html#setSoTimeout%28int%29) it will throw an `SocketTimeoutException` which will imediately terminate the whole exercise so we effectively broken the support for using this `RFXComTcpConnector`

martinvw commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComSerialConnector.java:
```diff
@@ -53,7 +53,7 @@ public void connect(RFXComBridgeConfiguration device)
         serialPort.setSerialPortParams(38400, SerialPort.DATABITS_8, SerialPort.STOPBITS_1, SerialPort.PARITY_NONE);
         serialPort.enableReceiveThreshold(1);
-        serialPort.disableReceiveTimeout();
+        serialPort.enableReceiveTimeout(100); // In ms. Small values mean faster shutdown but more cpu usage.
```
As we did discover this one does do what we hoped... http://www.docjava.com/book/cgij/jdoc/gnu/io/RXTXPort.html#enableReceiveTimeout(int)
